File: LISTAS_URI/lista5/dark_roads.py
from collections import defaultdict
import heapq
import sys

def prim(graph, source):
  weights = [sys.maxsize] * (max(graph) + 1)
  total_weight = 0
  queue = []
  visited = dict()

  weights[source] = 0
  queue.append((weights[source], source))

  while queue:
    heapq.heapify(queue)
    output = queue.pop(0)[1]
    visited[output] = True

    for adjacent in graph[output]:
      if adjacent not in visited and graph[output][adjacent] < weights[adjacent]:
        weights[adjacent] = graph[output][adjacent]
        total_weight += weights[adjacent-1]
        queue.append((weights[adjacent], adjacent))

  return weights

m, n = input().split(" ")
graph = defaultdict(dict)
total_cost = 0

while (int(m) != 0 and int(m) != 0):
  for i in range(int(n)):
    x, y, z = input().split(" ")
    graph[int(x)][int(y)] = int(z)
    graph[int(y)][int(x)] = int(z)
    total_cost += int(z)

  weights = prim(graph, min(graph))
  m, n = input().split(" ")
  print(total_cost - sum(weights))

File: LISTAS_URI/lista5/darkRoads.cpp
#include <iostream>
#include <vector>
#include <bits/stdc++.h>
#include <queue>

using namespace std;

typedef pair<int, int> node;

int prim(vector<node> graph[], int len_graph, int source) {
  int totalWeight = 0;
  priority_queue<node, vector<node>, greater<node>> queue;
  vector<int> weights(len_graph, INT_MAX);
  vector<bool> visited(len_graph, false);

  weights[source] = 0;
  queue.push(make_pair(weights[source], source));

  while (!queue.empty()) {
    int output = queue.top().second;
    queue.pop();
    visited[output] = true;

    for (vector<node>::iterator i = graph[output].begin(); i != graph[output].end(); i++) {
      if (visited[i->first] == false && i->second < weights[i->first]) {
        weights[i->first] = i->second;
        queue.push(make_pair(weights[i->first], i->first));
      }
    }
  }

  for (int i = 0; i < len_graph; i++) {
    totalWeight += weights[i];
  }

  return totalWeight;
}

int main(int argc, char **argv) {
  while (true) {
    int m, n;
    int totalWeight = 0;

    scanf("%d %d", &m, &n);

    if (m == 0 && n == 0) break;
    
    vector <pair<int, int>> graph[m];

    for (int i = 0; i < n; i++) {
      int x, y, z;

      scanf("%d %d %d", &x, &y, &z);
      graph[x].push_back(make_pair(y, z));
      graph[y].push_back(make_pair(x, z));
      totalWeight += z;
    }

    int mstTotalWeight = prim(graph, m, 0);
    printf("%d\n", totalWeight - mstTotalWeight);
  }

  return 0;
}
